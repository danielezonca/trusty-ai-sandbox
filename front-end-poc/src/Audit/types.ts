import { ExecutionType } from "../Shared/api/audit.api";

export interface IExecution {
  executionId: string;
  executionDate: string;
  executionType: string;
  executionSucceeded: boolean;
  executorName: string;
}

export interface IExecutionList {
  total: number;
  limit: number;
  offset: number;
  headers: IExecution[];
}

export interface IExecutionRouteParams {
  executionId: string;
  executionType: ExecutionType;
}
