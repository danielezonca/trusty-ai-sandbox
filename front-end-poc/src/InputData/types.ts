export interface IItemObject {
    name: string,
    typeRef: string,
    value: string | number | null,
    components: (IItemObject | IItemObject[])[],
    impact?: boolean | number,
    score?: number
}

export interface IInputRow {
    inputLabel: string,
    inputValue?: string | number | boolean | object | null,
    hasEffect?: boolean | number,
    score?: number,
    key: string,
    category: string
}

export function isIItemObjectArray(object: any): object is IItemObject[] {
    return typeof object[0].name == 'string';
}

export function isIItemObjectMultiArray(object: any): object is IItemObject[][] {
    return Array.isArray(object[0]);
}