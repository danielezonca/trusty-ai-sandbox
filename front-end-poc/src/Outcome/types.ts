import { IItemObject } from "../InputData/types";

export enum evaluationStatus {
  EVALUATING = "Evaluating",
  FAILED = "Failed",
  NOT_EVALUATED = "Not evaluated",
  SKIPPED = "skipped",
  SUCCEEDED = "Succeeded",
}

type evaluationStatusStrings = keyof typeof evaluationStatus;

export interface IOutcome {
  outcomeId: string;
  outcomeName: string;
  evaluationStatus: evaluationStatusStrings;
  hasErrors: boolean;
  messages: string[];
  outcomeResult: IItemObject;
}

export interface IOutcomeDetail extends IOutcome {
  outcomeInputs: IItemObject[];
}
