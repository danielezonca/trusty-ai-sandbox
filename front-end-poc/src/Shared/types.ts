export type RemoteData<E, D> =
  | { status: "NOT_ASKED" }
  | { status: "LOADING" }
  | { status: "FAILURE"; error: E }
  | { status: "SUCCESS"; data: D };
