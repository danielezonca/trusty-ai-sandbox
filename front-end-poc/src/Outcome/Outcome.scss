.outcomes-preview {
  .outcome {
    margin-bottom: var(--pf-global--spacer--md);
    &__title {
      margin-bottom: var(--pf-global--spacer--md);
    }
    &__property {
      margin-bottom: var(--pf-global--spacer--sm);
    }
  }
  .outcome .outcome {
    padding-left: var(--pf-global--spacer--md);
    border-left: 3px solid lightblue;
    .outcome {
      margin-top: var(--pf-global--spacer--md);
    }

  }
  & > .outcome-list {
    font-size: 1.125rem;
    padding: (10/18)+em 0;
    border-bottom: 1px solid lightgray;
    &:last-child {
      border-bottom: none;
    }
  }
}
.outcome-property {
  &__name {
    color: var(--pf-global--Color--200);
  }
}
.outcome-list {
  &__title {
    margin-bottom: var(--pf-global--spacer--md);
  }
  &__items {
    list-style: none;
  }
  &__item {
    margin-bottom: var(--pf-global--spacer--md);
    background-color: #f2f2f2;
    border-radius: 3px;
    padding: var(--pf-global--spacer--md);
    &__title {
      margin-bottom: var(--pf-global--spacer--lg);
      padding-bottom: var(--pf-global--spacer--sm);
      border-bottom: 1px solid lightgray;
    }
    .pf-l-grid__item {
      margin-bottom: var(--pf-global--spacer--sm);
    }
  }
}

.sticky-stuff {
  position: relative;
  &.is-sticky {
    overflow: hidden;
    .sticky-inner {
      position: fixed;
      top: 0;
      height: 100vh;
      overflow: auto;
    }
  }
}