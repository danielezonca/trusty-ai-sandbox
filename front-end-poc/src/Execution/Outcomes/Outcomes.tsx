import React from "react";
import { IOutcome } from "../../Outcome/types";
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  Gallery,
  GalleryItem,
  Split,
  SplitItem,
  Title,
} from "@patternfly/react-core";
import { v4 as uuid } from "uuid";
import EvaluationStatus from "../../Explanation/EvaluationStatus/EvaluationStatus";
import { LongArrowAltRightIcon } from "@patternfly/react-icons";
import { IItemObject, isIItemObjectArray, isIItemObjectMultiArray } from "../../InputData/types";
import "./Outcomes.scss";
import FormattedValue from "../../Shared/components/FormattedValue/FormattedValue";

type OutcomesProps = {
  outcomes: IOutcome[];
  onExplanationClick: (outcomeId: string) => void;
  listView?: boolean;
};

const Outcomes = (props: OutcomesProps) => {
  const { outcomes, onExplanationClick, listView = false } = props;

  if (listView) {
    return (
      <>
        {outcomes.length && (
          <Gallery className="outcome-cards" hasGutter>
            {outcomes.map((item) => renderCard(item, onExplanationClick))}
          </Gallery>
        )}
      </>
    );
  }

  return (
    <>
      {outcomes.map((item) => {
        if (
          item.outcomeResult !== null &&
          item.outcomeResult.components !== null &&
          isIItemObjectMultiArray(item.outcomeResult.components)
        ) {
          return item.outcomeResult.components.map((subList) => {
            return (
              <LightCard>
                <OutcomeSubList subListItem={subList} />
              </LightCard>
            );
          });
        }
        return <LightCard>{renderOutcome(item.outcomeResult, item.outcomeName, false, false)}</LightCard>;
      })}
    </>
  );
};

export default Outcomes;

const renderCard = (outcome: IOutcome, onExplanation: (outcomeId: string) => void) => {
  if (outcome.evaluationStatus !== "SUCCEEDED") {
    return (
      <GalleryItem key={uuid()}>
        <OutcomeCard outcome={outcome} onExplanation={onExplanation}>
          <span />
        </OutcomeCard>
      </GalleryItem>
    );
  }
  if (
    outcome.outcomeResult !== null &&
    outcome.outcomeResult.components !== null &&
    isIItemObjectMultiArray(outcome.outcomeResult.components)
  ) {
    return outcome.outcomeResult.components.map((item) => (
      <GalleryItem key={uuid()}>
        <OutcomeCard outcome={outcome} onExplanation={onExplanation}>
          <OutcomeSubList subListItem={item} />
        </OutcomeCard>
      </GalleryItem>
    ));
  }

  return (
    <GalleryItem key={uuid()}>
      <OutcomeCard outcome={outcome} onExplanation={onExplanation}>
        <div>{renderOutcome(outcome.outcomeResult, outcome.outcomeName, true, true)}</div>
      </OutcomeCard>
    </GalleryItem>
  );
};

type OutcomeCardProps = {
  children: React.ReactNode;
  outcome: IOutcome;
  onExplanation: (outcomeId: string) => void;
};

const OutcomeCard = (props: OutcomeCardProps) => {
  const { children, outcome, onExplanation } = props;
  return (
    <Card className="outcome-cards__card" isHoverable>
      <CardHeader>
        <Title headingLevel="h4" size="xl">
          {outcome.outcomeName}
        </Title>
      </CardHeader>
      <CardBody>
        {outcome.evaluationStatus !== undefined && outcome.evaluationStatus !== "SUCCEEDED" && (
          <div>
            <EvaluationStatus status={outcome.evaluationStatus} />
          </div>
        )}
        <div>{children}</div>
      </CardBody>
      <CardFooter>
        {outcome.outcomeId && onExplanation && (
          <Button
            variant="link"
            isInline
            className="outcome-cards__card__explanation-link"
            onClick={() => {
              onExplanation(outcome.outcomeId);
            }}>
            View Details <LongArrowAltRightIcon />
          </Button>
        )}
      </CardFooter>
    </Card>
  );
};

const renderOutcome = (outcomeData: IItemObject, name: string, compact = true, hidePropertyName = false) => {
  let renderItems: JSX.Element[] = [];

  if (outcomeData.value !== null) {
    return <OutcomeProperty property={outcomeData} key={outcomeData.name} hidePropertyName={hidePropertyName} />;
  }
  if (outcomeData.components.length) {
    if (isIItemObjectArray(outcomeData.components)) {
      renderItems.push(<OutcomeComposed outcome={outcomeData} key={outcomeData.name} compact={compact} name={name} />);
    } else if (isIItemObjectMultiArray(outcomeData.components)) {
      outcomeData.components.forEach((item) => {
        renderItems.push(<OutcomeSubList subListItem={item} />);
      });
    }
  }

  return <React.Fragment key={uuid()}>{renderItems.map((item: JSX.Element) => item)}</React.Fragment>;
};

const OutcomeProperty = (props: { property: IItemObject; hidePropertyName: boolean }) => {
  const { property, hidePropertyName } = props;
  const basicTypes = ["string", "number", "boolean"];
  const bigOutcome = hidePropertyName && basicTypes.includes(typeof property.value);

  if (bigOutcome)
    return (
      <div className="outcome__property__value--bigger">
        <FormattedValue value={property.value} />
      </div>
    );
  else
    return (
      <Split key={uuid()} className="outcome__property">
        <SplitItem className="outcome__property__name" key="property-name">
          {hidePropertyName ? "Result" : property.name}:
        </SplitItem>
        <SplitItem className="outcome__property__value" key="property-value">
          <FormattedValue value={property.value} />
        </SplitItem>
      </Split>
    );
};

const OutcomeComposed = (props: { outcome: IItemObject; compact: boolean; name: string }) => {
  const { outcome, compact, name } = props;
  let renderItems: JSX.Element[] = [];

  for (let subItem of outcome.components as IItemObject[]) {
    renderItems.push(
      <div className="outcome-item" key={subItem.name}>
        {renderOutcome(subItem, name, compact)}
      </div>
    );
  }
  return (
    <div className="outcome outcome--struct" key={outcome.name}>
      {renderItems.map((item) => item)}
    </div>
  );
};

type OutcomeSubListProps = {
  subListItem: IItemObject[];
};
const OutcomeSubList = (props: OutcomeSubListProps) => {
  const { subListItem } = props;

  return (
    <>
      {subListItem &&
        subListItem.map((item) => (
          <Split key={uuid()} className="outcome__property">
            <SplitItem className="outcome__property__name" key="property-name">
              {item.name}:
            </SplitItem>
            <SplitItem className="outcome__property__value" key="property-value">
              <FormattedValue value={item.value} />
            </SplitItem>
          </Split>
        ))}
    </>
  );
};

type LightCardProps = {
  children: React.ReactNode;
};

const LightCard = (props: LightCardProps) => {
  const { children } = props;
  return (
    <Card key={uuid()}>
      <CardBody>{children}</CardBody>
    </Card>
  );
};
