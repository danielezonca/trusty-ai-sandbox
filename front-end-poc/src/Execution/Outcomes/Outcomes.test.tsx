import React from "react";
import Outcomes from "./Outcomes";
import { mount } from "enzyme";
import { IOutcome } from "../../Outcome/types";

jest.mock("uuid", () => {
  let value = 0;
  return { v4: () => value++ };
});

const outcomesProps = {
  outcomes: [
    {
      outcomeId: "_12268B68-94A1-4960-B4C8-0B6071AFDE58",
      outcomeName: "Mortgage Approval",
      evaluationStatus: "SUCCEEDED",
      outcomeResult: { name: "Mortgage Approval", typeRef: "boolean", value: true, components: [] },
      messages: [],
      hasErrors: false,
    },
    {
      outcomeId: "_9CFF8C35-4EB3-451E-874C-DB27A5A424C0",
      outcomeName: "Risk Score",
      evaluationStatus: "SUCCEEDED",
      outcomeResult: { name: "Risk Score", typeRef: "number", value: 21.7031851958099, components: [] },
      messages: [],
      hasErrors: false,
    },
  ] as IOutcome[],
  onExplanationClick: jest.fn(),
  listView: true,
};

describe("Outcomes", () => {
  test("renders a list of outcomes as cards", () => {
    const wrapper = mount(<Outcomes {...outcomesProps} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("OutcomeCard")).toHaveLength(2);
    expect(wrapper.find("h4.outcome-cards__card__title").at(0).text()).toMatch("Mortgage Approval");
    expect(wrapper.find(".outcome__property__value--bigger").at(0).text()).toMatch("true");
    expect(wrapper.find("h4.outcome-cards__card__title").at(1).text()).toMatch("Risk Score");
    expect(wrapper.find(".outcome__property__value--bigger").at(1).text()).toMatch("21.7031851958099");
  });
  test("handles clicks on the explanation link", () => {
    const wrapper = mount(<Outcomes {...outcomesProps} />);

    wrapper.find(".outcome-cards__card__explanation-link").at(0).simulate("click");
    expect(outcomesProps.onExplanationClick).toHaveBeenCalledTimes(1);
    expect(outcomesProps.onExplanationClick).toHaveBeenCalledWith("_12268B68-94A1-4960-B4C8-0B6071AFDE58");
  });
});
