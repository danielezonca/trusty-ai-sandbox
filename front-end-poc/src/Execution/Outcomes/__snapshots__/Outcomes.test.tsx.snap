// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outcomes renders a list of outcomes as cards 1`] = `
<Outcomes
  listView={true}
  onExplanationClick={[MockFunction]}
  outcomes={
    Array [
      Object {
        "evaluationStatus": "SUCCEEDED",
        "hasErrors": false,
        "messages": Array [],
        "outcomeId": "_12268B68-94A1-4960-B4C8-0B6071AFDE58",
        "outcomeName": "Mortgage Approval",
        "outcomeResult": Object {
          "components": Array [],
          "name": "Mortgage Approval",
          "typeRef": "boolean",
          "value": true,
        },
      },
      Object {
        "evaluationStatus": "SUCCEEDED",
        "hasErrors": false,
        "messages": Array [],
        "outcomeId": "_9CFF8C35-4EB3-451E-874C-DB27A5A424C0",
        "outcomeName": "Risk Score",
        "outcomeResult": Object {
          "components": Array [],
          "name": "Risk Score",
          "typeRef": "number",
          "value": 21.7031851958099,
        },
      },
    ]
  }
>
  <section
    className="outcomes"
  >
    <Component
      className="outcome-cards"
      hasGutter={true}
    >
      <div
        className="pf-l-gallery pf-m-gutter outcome-cards"
      >
        <Component
          key="0"
        >
          <div>
            <OutcomeCard
              onExplanation={[MockFunction]}
              outcome={
                Object {
                  "evaluationStatus": "SUCCEEDED",
                  "hasErrors": false,
                  "messages": Array [],
                  "outcomeId": "_12268B68-94A1-4960-B4C8-0B6071AFDE58",
                  "outcomeName": "Mortgage Approval",
                  "outcomeResult": Object {
                    "components": Array [],
                    "name": "Mortgage Approval",
                    "typeRef": "boolean",
                    "value": true,
                  },
                }
              }
            >
              <Component
                className="outcome-cards__card"
                isHoverable={true}
              >
                <article
                  className="pf-c-card pf-m-hoverable outcome-cards__card"
                >
                  <Component>
                    <div
                      className="pf-c-card__header"
                    >
                      <Component
                        className="outcome-cards__card__title"
                        headingLevel="h4"
                        size="xl"
                      >
                        <h4
                          className="pf-c-title pf-m-xl outcome-cards__card__title"
                        >
                          Mortgage Approval
                        </h4>
                      </Component>
                    </div>
                  </Component>
                  <Component>
                    <div
                      className="pf-c-card__body"
                    >
                      <div>
                        <div>
                          <OutcomeProperty
                            hidePropertyName={true}
                            key="Mortgage Approval"
                            property={
                              Object {
                                "components": Array [],
                                "name": "Mortgage Approval",
                                "typeRef": "boolean",
                                "value": true,
                              }
                            }
                          >
                            <div
                              className="outcome__property__value--bigger"
                            >
                              <FormattedValue
                                value={true}
                              >
                                <span
                                  className="formatted-value formatted-value--capitalize"
                                >
                                  true
                                </span>
                              </FormattedValue>
                            </div>
                          </OutcomeProperty>
                        </div>
                      </div>
                    </div>
                  </Component>
                  <Component>
                    <div
                      className="pf-c-card__footer"
                    >
                      <Component
                        className="outcome-cards__card__explanation-link"
                        isInline={true}
                        onClick={[Function]}
                        variant="link"
                      >
                        <button
                          aria-disabled={null}
                          aria-label={null}
                          className="pf-c-button pf-m-link pf-m-inline outcome-cards__card__explanation-link"
                          data-ouia-component-id={null}
                          data-ouia-component-type="PF4/Button"
                          data-ouia-safe={true}
                          disabled={false}
                          onClick={[Function]}
                          tabIndex={null}
                          type="button"
                        >
                          View Details 
                          <LongArrowAltRightIcon
                            color="currentColor"
                            noVerticalAlign={false}
                            size="sm"
                          >
                            <svg
                              aria-hidden={true}
                              aria-labelledby={null}
                              fill="currentColor"
                              height="1em"
                              role="img"
                              style={
                                Object {
                                  "verticalAlign": "-0.125em",
                                }
                              }
                              viewBox="0 0 448 512"
                              width="1em"
                            >
                              <path
                                d="M313.941 216H12c-6.627 0-12 5.373-12 12v56c0 6.627 5.373 12 12 12h301.941v46.059c0 21.382 25.851 32.09 40.971 16.971l86.059-86.059c9.373-9.373 9.373-24.569 0-33.941l-86.059-86.059c-15.119-15.119-40.971-4.411-40.971 16.971V216z"
                                transform=""
                              />
                            </svg>
                          </LongArrowAltRightIcon>
                        </button>
                      </Component>
                    </div>
                  </Component>
                </article>
              </Component>
            </OutcomeCard>
          </div>
        </Component>
        <Component
          key="1"
        >
          <div>
            <OutcomeCard
              onExplanation={[MockFunction]}
              outcome={
                Object {
                  "evaluationStatus": "SUCCEEDED",
                  "hasErrors": false,
                  "messages": Array [],
                  "outcomeId": "_9CFF8C35-4EB3-451E-874C-DB27A5A424C0",
                  "outcomeName": "Risk Score",
                  "outcomeResult": Object {
                    "components": Array [],
                    "name": "Risk Score",
                    "typeRef": "number",
                    "value": 21.7031851958099,
                  },
                }
              }
            >
              <Component
                className="outcome-cards__card"
                isHoverable={true}
              >
                <article
                  className="pf-c-card pf-m-hoverable outcome-cards__card"
                >
                  <Component>
                    <div
                      className="pf-c-card__header"
                    >
                      <Component
                        className="outcome-cards__card__title"
                        headingLevel="h4"
                        size="xl"
                      >
                        <h4
                          className="pf-c-title pf-m-xl outcome-cards__card__title"
                        >
                          Risk Score
                        </h4>
                      </Component>
                    </div>
                  </Component>
                  <Component>
                    <div
                      className="pf-c-card__body"
                    >
                      <div>
                        <div>
                          <OutcomeProperty
                            hidePropertyName={true}
                            key="Risk Score"
                            property={
                              Object {
                                "components": Array [],
                                "name": "Risk Score",
                                "typeRef": "number",
                                "value": 21.7031851958099,
                              }
                            }
                          >
                            <div
                              className="outcome__property__value--bigger"
                            >
                              <FormattedValue
                                value={21.7031851958099}
                              >
                                <span
                                  className="formatted-value"
                                >
                                  21.7031851958099
                                </span>
                              </FormattedValue>
                            </div>
                          </OutcomeProperty>
                        </div>
                      </div>
                    </div>
                  </Component>
                  <Component>
                    <div
                      className="pf-c-card__footer"
                    >
                      <Component
                        className="outcome-cards__card__explanation-link"
                        isInline={true}
                        onClick={[Function]}
                        variant="link"
                      >
                        <button
                          aria-disabled={null}
                          aria-label={null}
                          className="pf-c-button pf-m-link pf-m-inline outcome-cards__card__explanation-link"
                          data-ouia-component-id={null}
                          data-ouia-component-type="PF4/Button"
                          data-ouia-safe={true}
                          disabled={false}
                          onClick={[Function]}
                          tabIndex={null}
                          type="button"
                        >
                          View Details 
                          <LongArrowAltRightIcon
                            color="currentColor"
                            noVerticalAlign={false}
                            size="sm"
                          >
                            <svg
                              aria-hidden={true}
                              aria-labelledby={null}
                              fill="currentColor"
                              height="1em"
                              role="img"
                              style={
                                Object {
                                  "verticalAlign": "-0.125em",
                                }
                              }
                              viewBox="0 0 448 512"
                              width="1em"
                            >
                              <path
                                d="M313.941 216H12c-6.627 0-12 5.373-12 12v56c0 6.627 5.373 12 12 12h301.941v46.059c0 21.382 25.851 32.09 40.971 16.971l86.059-86.059c9.373-9.373 9.373-24.569 0-33.941l-86.059-86.059c-15.119-15.119-40.971-4.411-40.971 16.971V216z"
                                transform=""
                              />
                            </svg>
                          </LongArrowAltRightIcon>
                        </button>
                      </Component>
                    </div>
                  </Component>
                </article>
              </Component>
            </OutcomeCard>
          </div>
        </Component>
      </div>
    </Component>
  </section>
</Outcomes>
`;
